### PR TITLE
Issue #9003: nondex tool found issue in token ordering at Indentation

### LINKED ISSUE #9003: nondex tool found issue in token ordering at Indentation
Identified at https://travis-ci.org/github/checkstyle/checkstyle/jobs/675893545#L3230 and resulted from merging https://github.com/checkstyle/checkstyle/pull/8099 it was reverted and second PR (https://github.com/checkstyle/checkstyle/pull/8140) stuck in blocking state,
initial try to fix was done at https://github.com/checkstyle/checkstyle/pull/8116

I believe the issue is from IndentationCheck and that it generates its list of tokens using a Set, which doesn't guarantee order. Nondex runs could clearly be seen randomizing the list of tokens, so I fixed it by sorting the tokens.

How to debug nondex issues (This is not by all means a detailed guide):

I found that right after the test failures, nondex reports commands that can be used to directly re-run the failure for debugging.
Example: https://travis-ci.org/github/checkstyle/checkstyle/jobs/675893545#L3237-L3238
> `mvn nondex:nondex  -DnondexFilter='.*' -DnondexMode=FULL -DnondexSeed=933178 -DnondexStart=0 -DnondexEnd=9223372036854775807 -DnondexPrintstack=false -DnondexDir="/home/travis/build/checkstyle/checkstyle/.nondex" -DnondexJarDir="/home/travis/build/checkstyle/checkstyle/.nondex" -DnondexExecid=KSFBHLJqhU3VzpVTwCrofpz33iuNv66h8ERN4hf2Zg= -DnondexLogging=CONFIG`

I used this command and added many console print outs to determine how the check was failing and compared it to the normal run without nondex. This is what lead me to the sorting issue. Since many runs always pointed back to IndentationCheck, I could make the deduction that it is because of the Set IndentationCheck uses to build its list of tokens.

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #9003: nondex tool found issue in token ordering at Indentation

This PR hopefully resolves #9003.  I have included the nondex update to verify that using an ordered data structure (LinkedHashSet) vs. a non-ordered structure (HashSet) resolves the issue.

First CI run:
failure at https://checkstyle.semaphoreci.com/jobs/d793aa43-1d5e-4058-991f-432f75525a99
image

`AllChecksTest.testAllCheckTokensAreReferencedInGoogleConfigFile:394->validateAllCheckTokensAreReferencedInConfigFile:451->validateDefaultTokens:479 All default tokens should be used in config for Indentation`

` JavadocPropertiesGeneratorTest.testNonExistentArgument:111 Unexpected error log ==> expected: <Missing required options and parameters: '--destfile=<outputFile>', '<inputFile>'`

### CODE REVIEW

**Comment by romani** [comment]
Nondex in semaphore is red

**Comment by nmancus1** [comment]
After running nondex's flaky test tool, I have found the following failures:
https://nmancus1.github.io/nondex/DHYOT7wzFcv1D5HlkdDcLF8e9PRvUQdmKaDrsD5Kw=/test_results.html

The `.txt` files with specific errors that correspond to each failing test can be found here:
https://nmancus1.github.io/nondex/DHYOT7wzFcv1D5HlkdDcLF8e9PRvUQdmKaDrsD5Kw=/outputfile.html

Checkstyle has over 300 failing tests, many for `WeakHashMap` violations, from using `Set`s instead of an ordered data structure...  I have a feeling that these will keep appearing sporadically unless they are "fixed".  How should we proceed?

**Comment by rnveach** [comment]
> https://nmancus1.github.io/nondex/DHYOT7wzFcv1D5HlkdDcLF8e9PRvUQdmKaDrsD5Kw=/com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyForInitializerPadCheckTest.txt
> java.lang.NoClassDefFoundError: java/util/WeakHashMap$HashIterator$HashIteratorShuffler

I don't understand how no class defined errors are coming up. Is this a bug in the tool?

**Comment by nmancus1** [comment]
@rnveach 
```
➜  ~ archlinux-java status
Available Java environments:
  java-10-openjdk
  java-11-openjdk
  java-14-openjdk
  java-15-jdk
  java-8-openjdk (default)
  jetbrains-jre
➜  ~ java -version
openjdk version "1.8.0_265"
OpenJDK Runtime Environment (build 1.8.0_265-b01)
OpenJDK 64-Bit Server VM (build 25.265-b01, mixed mode)

```
```
➜  checkstyle git:(issue-9003-set-in-indentation) ✗ cat pom.xml | grep surefire
        <artifactId>maven-surefire-plugin</artifactId>
        <artifactId>maven-surefire-report-plugin</artifactId>
            <excludedPage>surefire-report.html</excludedPage>

```
From nondex github repo:
```
Prerequisites:
- Java 8 (Oracle JDK, OpenJDK).
- Surefire present in the POM.
```
The documentation for the nondex tool is not very helpful; I can't find anything other than one paper and the README.md on their github.  I thought that all of the methods/ files reported were violations since the ones that I checked did use unsorted data structures, usually sets, and many iterated through the sets using an enhanced for loop, thus triggering the "assumption on an under-determined Java API" violations. 

>This means some class is missing, but it is coming from `java.util.` package.

This is correct, I initially thought this was nondex's particular method for reporting violations, but in hindsight that doesn't make much sense. This must be a bug in nondex, since my local machine seems to meet the prerequisites to use the tool.

**Comment by nmancus1** [comment]
Ok, I've fixed the 

`AllChecksTest.testAllCheckTokensAreReferencedInGoogleConfigFile:394->validateAllCheckTokensAreReferencedInConfigFile:451->validateDefaultTokens:479 All default tokens should be used in config for Indentation` 

issue,  I am going to push my changes to verify in CI, then work on the 

`JavadocPropertiesGeneratorTest.testNonExistentArgument:111 Unexpected error log ==> expected: <Missing required options and parameters: '--destfile=<outputFile>', '<inputFile>'` 

failure

**Comment by rnveach** [comment]
Thanks @nmancus1 . If you haven't already, I would report this as a bug in a new issue on their repo, especially if you can reproduce it on multiple runs. The exception makes no sense but it looks to closely resemble https://github.com/TestingResearchIllinois/NonDex/issues/98 .

**Comment by nmancus1** [comment]
Nondex is failing to produce debugging information for the last failing test, `JavadocPropertiesGeneratorTest`, and a few attempts to guess what could be the issue have failed.  I have added `JavadocPropertiesGeneratorTest` to the `excludedGroups` section of the nondex configuration in `pom.xml`, until the issue I have opened at https://github.com/TestingResearchIllinois/NonDex/issues/161 is resolved.

**Comment by rnveach** [comment]
@nmancus1 If you are waiting on something from us besides review, please let us know what it is. If this PR is ready to be finalized, please rebase it on the latest master.

File: .ci/validation.sh [hunk 1]
```diff
@@ -88,7 +88,8 @@ eclipse-static-analysis)
   ;;
 
 nondex)
-  mvn -e --fail-never clean nondex:nondex -DargLine='-Xms1024m -Xmx2048m'
+  mvn -e --fail-never clean nondex:nondex -DargLine='-Xms1024m -Xmx2048m' \
+    -Dtest=!JavadocPropertiesGeneratorTest#testNonExistentArgument
```
Comment (nmancus1) [inline .ci/validation.sh:92]: done to resolve ` JavadocPropertiesGeneratorTest.testNonExistentArgument:111 Unexpected error log ==> expected: <Missing required options and parameters: '--destfile=<outputFile>', '<inputFile>'08:59
Usage: java com.puppycrawl.tools.checkstyle.JavadocPropertiesGenerator [-hV]08:59
       --destfile=<outputFile> <inputFile>08:59
      <inputFile>   The input file.08:59
      --destfile=<outputFile>08:59
                    The output file.08:59
  -h, --help        Show this help message and exit.08:59
  -V, --version     Print version information and exit.08:59
> but was: <Missing required parameter: '<inputFile>'08:59
Usage: java com.puppycrawl.tools.checkstyle.JavadocPropertiesGenerator [-hV]08:59
       --destfile=<outputFile> <inputFile>08:59
      <inputFile>   The input file.08:59
      --destfile=<outputFile>08:59
                    The output file.08:59
  -h, --help        Show this help message and exit.08:59
  -V, --version     Print version information and exit.08:59
>`

File: src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java [hunk 2]
```diff
@@ -280,7 +280,7 @@ public void testHasIntersectionEarlyOut() throws Exception {
                 "clangComments");
         final TextBlock textBlock = new Comment(new String[] {""}, 1, 1, 1);
         clangComments.put(1, Collections.singletonList(textBlock));
-        clangComments.put(2, null);
+        clangComments.put(2, Collections.emptyList());
```
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java:283]: resolves `FileContentsTest.testHasIntersectionEarlyOut:285 » NullPointer`

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java [hunk 3]
```diff
@@ -471,7 +473,16 @@ private static boolean isAllTokensAcceptable(AbstractCheck check) {
...
+
+        final Set<Integer> defaultTokensSet = IntStream.of(check.getDefaultTokens())
+            .boxed()
+            .collect(Collectors.toSet());
+
+        final Set<Integer> requiredTokensSet = IntStream.of(check.getRequiredTokens())
+            .boxed()
+            .collect(Collectors.toSet());
+
+        if (defaultTokensSet.equals(requiredTokensSet)) {
```
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:485]: all changes in this file resolve `AllChecksTest.testAllCheckTokensAreReferencedInGoogleConfigFile:394->validateAllCheckTokensAreReferencedInConfigFile:451->validateDefaultTokens:479 All default tokens should be used in config for Indentation`

File: .ci/validation.sh (hunk 1 above)
Comment (nmancus1) [reply .ci/validation.sh:92]: This error was likely caused by the picocli library; this is probably why the nondex report failed to generate a file for this.

Comment (rnveach) [reply .ci/validation.sh:92]: Please add this reason as a comment to the script file.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java [hunk 4]
```diff
@@ -78,9 +80,9 @@
     };
 
     private static final Map<String, Set<String>> CHECKSTYLE_TOKENS_IN_CONFIG_TO_IGNORE =
-            new HashMap<>();
+            new LinkedHashMap<>();
     private static final Map<String, Set<String>> GOOGLE_TOKENS_IN_CONFIG_TO_IGNORE =
-            new HashMap<>();
+            new LinkedHashMap<>();
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:85]: If the other changes resolve https://github.com/checkstyle/checkstyle/pull/9083#discussion_r553663124 , then why are these changes needed?

**Review by rnveach — COMMENTED** [review]
1 question.

File: src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java (hunk 2 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java:283]: This now causes a pitest failure.
https://app.circleci.com/pipelines/github/checkstyle/checkstyle/4214/workflows/227adf91-df2d-4285-8fda-f24cbd7c8afa/jobs/69854
````
+FileContents.java.html:<td class='covered'><pre><span  class='survived'>            if (hasIntersection) {</span></pre></td></tr>
````

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java (hunk 4 above)
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: They are not, thanks for catching that. Fixed.

File: .ci/validation.sh (hunk 1 above)
Comment (nmancus1) [reply .ci/validation.sh:93]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java (hunk 2 above)
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java:283]: Fixed.

**Review by rnveach — APPROVED** [review]

**Review by romani — APPROVED** [review]
